aesara/tensor/subtensor_opt.py

```diff
@@ -879,135 +879,142 @@ def local_set_to_inc_subtensor(fgraph, node):
+                pass
 
-            if start != 0:
+            if idx.stop < length_pos_data:
                 return False
-            if stop != length:
+        elif isinstance(idx.stop, Variable):
+            length_pos_shape_i = idx.stop
+            # length_pos is a tensor variable, but length_pos_shape_i
+            # is a scalar variable. We try to see if they represent
+            # the same underlying variable.
+            if length_pos_shape_i.owner and isinstance(
+                length_pos_shape_i.owner.op, ScalarFromTensor
+            ):
+                length_pos_shape_i = length_pos_shape_i.owner.inputs[0]
+            elif length_pos.owner and isinstance(length_pos.owner.op, TensorFromScalar):
+                length_pos = length_pos.owner.inputs[0]
+            else:
+                # We did not find underlying variables of the same type
                 return False
-            if step != 1:
+
+            # The type can be different: int32 vs int64. length_pos
+            # should always be int64 as that is what the shape
+            # tracker keep. Subtensor accept any scalar int{8,16,32,64}
+            # as index type.
+            assert str(length_pos.type.dtype) == "int64"
+            assert str(length_pos_shape_i.type.dtype) in [
+                "int8",
+                "int16",
+                "int32",
+                "int64",
+            ]
+
+            # length_pos_shape_i cannot be None
+            if length_pos_shape_i != length_pos:
                 return False
+        elif idx.stop is None:
+            continue
         else:
             return False
+
+    return [node.inputs[0]]
+
+
+@register_canonicalize
+@register_specialize
+@local_optimizer([AdvancedSubtensor1])
+def local_useless_AdvancedSubtensor1(fgraph, node):
+    """Remove `AdvancedSubtensor1` if it takes the full input.
+
+    In the `AdvancedSubtensor1` case, the full input is taken when the indices
+    are equivalent to ``arange(0, input.shape[0], 1)`` using either an explicit
+    list/vector or the `ARange` `Op`.
+
+    """
+    # This optimization needs ShapeOpt and fgraph.shape_feature
+    if not hasattr(fgraph, "shape_feature"):
+        return
+
+    shape_of = fgraph.shape_feature.shape_of
+
+    # get length of the indexed tensor along the first axis
+    try:
+        length = get_scalar_constant_value(
+            shape_of[node.inputs[0]][0], only_process_constants=True
+        )
+    except NotScalarConstantError:
+        return False
+
+    # get index (which must be a vector by definition)
+    idx = node.inputs[1]
+
+    # `idx` must be equivalent to [0,1,...,shape[0] - 1] to qualify for
+    # this optimization
+    if isinstance(idx, Constant):
+        idx = idx.value
+        if len(idx) != length:
+            return False
+        if np.any(idx != np.arange(length)):
+            return False
+    elif idx.owner is not None and isinstance(idx.owner.op, ARange):
+        try:
```

Comment:
A lot of missing coverage, I guess that was already the case before?

Reply:
Yes, unfortunately.

Reply:
Now that I think about it, that whole `ARange` branch is effectively unreachable and/or pointless, because, if `idx.owner.inputs` are all constants (as the rewrite requires), the entire node would've been constant folded before reaching that point (i.e. there would be no `ARange` node at all).